### ISSUE TITLE
regression 3.1.0: tsql: variable assignment is incorrectly fixed and breaks code. rule  CV05

### ISSUE THREAD

**Issue by fmms** [sqlfluff/sqlfluff#6000]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I have updated to sqlfluff 3.1.0.

I wanted to format:
```sql
DECLARE @job_notify_netsend_operator_name bigint;
SELECT @job_notify_netsend_operator_name = NULL;
```

### Expected Behaviour

code is left untouched.

### Observed Behaviour

code is fixed to

```sql
DECLARE @job_notify_netsend_operator_name bigint;
SELECT @job_notify_netsend_operator_name IS NULL;
```

and thus totally broken.

### How to reproduce

see above

```
sqlfluff fix .\query1.sql
==== finding fixable violations ====
== [query1.sql] FAIL
L:   2 | P:  42 | CV05 | Comparisons with NULL should use "IS" or "IS NOT".
                       | [convention.is_null]
== [query1.sql] FIXED
1 fixable linting violations found
```

### Dialect

tsql

### Version

3.1.0

### Configuration

none

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by drjwelch** [comment]
I have found the reason: in 3.0.7 this rule is not checked because your statement is incorrectly parsed; now I fixed the parsing, the rule does run and is violated because it doesn't expect the T-SQL syntax `SELECT @param=value` (it does allow the preferred syntax, `SET @param = value`)

Workaround: for now (and it's the preferred syntax anyway) change `SELECT @param=value` to `SET @param=value` in your code

I'll work on a fix for this: rule CV05 for T-SQL dialect should (arguably) allow your version.

Detailed reason:

Sample SQL: `SELECT @a = NULL`  

Parse tree in 3.0.7: note the incorrect parsing as `alias_expression` and missing `comparison_operator` child node
```
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    batch:
[L:  1, P:  1]      |        statement:
[L:  1, P:  1]      |            select_statement:
[L:  1, P:  1]      |                select_clause:
[L:  1, P:  1]      |                    keyword:                                  'select'
[L:  1, P:  7]      |                    [META] indent:
[L:  1, P:  7]      |                    whitespace:                               ' '
[L:  1, P:  8]      |                    select_clause_element:
[L:  1, P:  8]      |                        alias_expression:
[L:  1, P:  8]      |                            parameter:                        '@a'
[L:  1, P: 10]      |                            whitespace:                       ' '
[L:  1, P: 11]      |                            raw_comparison_operator:          '='
[L:  1, P: 12]      |                        whitespace:                           ' '
[L:  1, P: 13]      |                        null_literal:                         'null'
[L:  1, P: 17]      |                    [META] dedent:
[L:  1, P: 17]      |    newline:                                                  '\n'
[L:  2, P:  1]      |    [META] end_of_file:
```
Parse tree in 3.1.0: note presence of `comparison_operator` child node
```
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    batch:
[L:  1, P:  1]      |        statement:
[L:  1, P:  1]      |            select_statement:
[L:  1, P:  1]      |                select_clause:
[L:  1, P:  1]      |                    keyword:                                  'select'
[L:  1, P:  7]      |                    [META] indent:
[L:  1, P:  7]      |                    whitespace:                               ' '
[L:  1, P:  8]      |                    select_clause_element:
[L:  1, P:  8]      |                        expression:
[L:  1, P:  8]      |                            parameter:                        '@a'
[L:  1, P: 10]      |                            whitespace:                       ' '
[L:  1, P: 11]      |                            comparison_operator:
[L:  1, P: 11]      |                                raw_comparison_operator:      '='
[L:  1, P: 12]      |                            whitespace:                       ' '
[L:  1, P: 13]      |                            null_literal:                     'null'
[L:  1, P: 17]      |                    [META] dedent:
[L:  1, P: 17]      |    newline:                                                  '\n'
[L:  2, P:  1]      |    [META] end_of_file:
```

Rule CV05 only runs if the parse tree contains `comparison_operator` child node: so it didn't run on your code in 3.0.7, which was a bug, but it does run on your code in 3.1.0.

The rule excludes the case where the comparison operator's parent is SET ... I will see if I can work out how to also exclude your case, for TSQL.

**Comment by drjwelch** [comment]
I think I've resolved this issue for T-SQL by abstracting-out the parameter assignment segment as I mentioned above.

I've added test cases and a new T-SQL-only rule to fail parameter assignment under SELECT and prefer SET.  I haven't managed to offer a lint fix yet (but the rule could be set to be ignored in the meantime).

I'm not available this week but I'll raise a PR next weekend for review.

**Comment by peterbud** [comment]
This seems to be working with the latest `sqlfluff`. I think #7134 has fixed it. Pls close the issue.

**Comment by drjwelch** [comment]
@peterbud I would describe that change as a workaround since it doesn't fix the root cause grammar issue (assignment classified as comparison), instead it modifies the test CV05.  I'd still like to merge my changes above but I might need some help in solving my test fixture issue.

**Comment by peterbud** [comment]
Hi @drjwelch , I think the root cause has been fixed. Maybe not just with #7134  but also with #7222.

Fact is, if you use the above mentioned example, it works as expected:

Input:
```sql
DECLARE @job_notify_netsend_operator_name bigint;
SELECT @job_notify_netsend_operator_name = NULL;

```

```bash
sqlfluff fix test.sql  
==== finding fixable violations ====
==== no fixable linting violations found ====                                                                                                                                           
All Finished 📜 🎉!
```

And the output is the same as the input.

also the parse tree is properly recognizing that this is a variable assignment (`select_variable_assignment`):
```bash
sqlfluff parse test.sql
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    batch:
[L:  1, P:  1]      |        statement:
[L:  1, P:  1]      |            declare_segment:
[L:  1, P:  1]      |                keyword:                                      'DECLARE'
[L:  1, P:  8]      |                [META] indent:
[L:  1, P:  8]      |                whitespace:                                   ' '
[L:  1, P:  9]      |                parameter:                                    '@job_notify_netsend_operator_name'
[L:  1, P: 42]      |                whitespace:                                   ' '
[L:  1, P: 43]      |                data_type:
[L:  1, P: 43]      |                    keyword:                                  'bigint'
[L:  1, P: 49]      |                [META] dedent:
[L:  1, P: 49]      |        statement_terminator:                                 ';'
[L:  1, P: 50]      |        newline:                                              '\n'
[L:  2, P:  1]      |        statement:
[L:  2, P:  1]      |            select_statement:
[L:  2, P:  1]      |                select_clause:
[L:  2, P:  1]      |                    keyword:                                  'SELECT'
[L:  2, P:  7]      |                    [META] indent:
[L:  2, P:  7]      |                    whitespace:                               ' '
[L:  2, P:  8]      |                    select_clause_element:
[L:  2, P:  8]      |                        select_variable_assignment:
[L:  2, P:  8]      |                            parameter:                        '@job_notify_netsend_operator_name'
[L:  2, P: 41]      |                            whitespace:                       ' '
[L:  2, P: 42]      |                            assignment_operator:
[L:  2, P: 42]      |                                raw_comparison_operator:      '='
[L:  2, P: 43]      |                            whitespace:                       ' '
[L:  2, P: 44]      |                            expression:
[L:  2, P: 44]      |                                null_literal:                 'NULL'
[L:  2, P: 48]      |                    [META] dedent:
[L:  2, P: 48]      |        statement_terminator:                                 ';'
[L:  2, P: 49]      |    newline:                                                  '\n'
[L:  3, P:  1]      |    [META] end_of_file:
```

Let me know if you have more questions.

**Comment by drjwelch** [comment]
Yes - your #7222 reflects everything I did above - and in fact you've backed-out #7134 as redundant now the parsing is fixed, so that one's kinda irrelevant.  Agree this should be closed now.
